formats/protobuf-generator/commonMain/src/kotlinx.serialization.protobuf.generator/Generation.kt

```diff
@@ -0,0 +1,383 @@
+                addCustomTypeWithElements(childDescriptor, all)
+            }
+        }
+        descriptor.isProtobufOpenMessage -> {
+            if (!all.containsKey(descriptor.serialName)) {
+                all[descriptor.serialName] = descriptor
+            }
+        }
+        descriptor.isProtobufContextualMessage -> return
+        else -> throw IllegalStateException(
+            "Unrecognized custom type with serial name "
+                    + "'${descriptor.serialName}' and kind '${descriptor.kind}'"
+        )
+    }
+}
+
+private fun generateMessage(
+    message: CustomTypeDeclaration,
+    builder: StringBuilder
+) {
+    val serialDescriptor = message.descriptor
+
+    builder.append("// serial name '").append(removeLineBreaks(serialDescriptor.serialName)).appendLine('\'')
+
+    builder.append("message ").append(message.name).appendLine(" {")
+    for (index in 0 until serialDescriptor.elementsCount) {
+        val childDescriptor = serialDescriptor.getElementDescriptor(index)
+        val annotations = serialDescriptor.getElementAnnotations(index)
+
+        val originFieldName = serialDescriptor.getElementName(index)
+        val fieldName = makeProtobufIdent(originFieldName)
+
+        if (originFieldName != fieldName) builder.append("  // original field name '")
+            .append(removeLineBreaks(originFieldName))
+            .appendLine('\'')
+
+        if (serialDescriptor.isElementOptional(index)) {
+            builder.appendLine("  // WARNING: field has a default value that is not present in the scheme")
+            println(
+                """WARNING: field '$fieldName' in serializable class '${serialDescriptor.serialName}' """ +
+                        "has a default value that is not present in the scheme!"
+            )
+        }
+
+        try {
+            when {
+                childDescriptor.isProtobufNamedType -> generateNamedType(
+                    serialDescriptor,
+                    childDescriptor,
+                    index,
+                    builder
+                )
+                childDescriptor.isProtobufMap -> generateMapType(childDescriptor, builder)
+                childDescriptor.isProtobufRepeated -> generateListType(childDescriptor, builder)
+                else -> throw IllegalStateException(
+                    "Unprocessed message field type with serial name " +
+                            "'${childDescriptor.serialName}' and kind '${childDescriptor.kind}'"
+                )
+            }
+        } catch (e: IllegalArgumentException) {
+            throw IllegalArgumentException(
+                "An error occurred during value generation for field $fieldName of message ${serialDescriptor.protobufCustomTypeName} " +
+                        "(serial name ${serialDescriptor.serialName}): ${e.message}", e
+            )
+        } catch (e: Exception) {
+            throw IllegalStateException(
```

Comment:
I'd add a comment that this branch is for internal errors, and previous is for incorrect user input.